### ISSUE TITLE
Early Stopping/Save Best Val Weights?

### ISSUE THREAD

**Issue by jmhessel** [IndicoDataSolutions/finetune#150]
Thanks for writing this library to wrap the OpenAI code!! I don't think the current codebase does this (I was looking in the base model's _train_loop and validation_hook) but I could be wrong.

**Is your feature request related to a problem? Please describe.**

When I am fitting the classifier, I would like to save the weights of the model with the best validation data loss. In particular -- I call:

```
model = finetune.Classifier(n_epochs=10,
                            test_size=.3,
                            verbose=True,
                            batch_size=8,
                            lm_loss_coef=0,
                            val_interval=25)
model.fit(train_data, train_labels)
```

I watch as the validation loss goes down, and then goes back up again...

```
...
Train loss: 0.8263243668572684   Validation loss: 0.9696971137060008
Train loss: 0.7893467535291913   Validation loss: 0.9315027973123593
Train loss: 0.7428553240288371   Validation loss: 0.8978088509162531
Train loss: 0.6532491042988868   Validation loss: 0.8657591236030956
Train loss: 0.5872814164979386   Validation loss: 0.8302381457714927
Train loss: 0.5188518893581162   Validation loss: 0.8055230206469525
Train loss: 0.46836010163226455  Validation loss: 0.7790362483260943
Train loss: 0.42655090732702466  Validation loss: 0.7963167028100855
Train loss: 0.37630249256415615  Validation loss: 0.7668756079191992
Train loss: 0.3127427928071437   Validation loss: 0.7535473302166583
Train loss: 0.26105406358681316  Validation loss: 0.752195528951159
Train loss: 0.2315318344735824   Validation loss: 0.74845844850349
Train loss: 0.20082835943056582  Validation loss: 0.7339546502212082
Train loss: 0.16591417155263366  Validation loss: 0.7224668118229861
Train loss: 0.16113676263676516  Validation loss: 0.71280302448007
Train loss: 0.1351519643111266   Validation loss: 0.7155087849984496
Train loss: 0.1122367973046417   Validation loss: 0.7097793683152129
Train loss: 0.08857986326043485  Validation loss: 0.7209393355041678
Train loss: 0.07383205768643403  Validation loss: 0.7290630419633354
Train loss: 0.05744698858779037  Validation loss: 0.737265006848874
Train loss: 0.04475240544373843  Validation loss: 0.7473114921713067
Train loss: 0.03484421883801572  Validation loss: 0.7718343255340118
...
```
Also -- the point at which the minimum is reached varies run-to-run, so I can't simply set the number of epochs apriori.

**Describe the solution you'd like**
It would be great to have an option to cache the temporary weights and then, at the end of the training loop, have the best weights according to validation loss reloaded automatically.

**Describe alternatives you've considered**
I could write my own training loop where I call fit (epochs=1) and save, but caching the weights and then reloading the best weights according to in a single call would be extremely convenient.

**Comment by madisonmay** [comment]
Hi @jmhessel,

We do have support for saving intermediate best model state as the model trains via the poorly documented [`autosave_path` setting.](https://finetune.indico.io/#finetune.config.Settings), but you're correct that we don't have actual early stopping implemented. However, we're in the middle of a large refactor to move to using the `tf.Estimator` framework, so this is a feature we should effectively get for free after this refactor is completed.  

I'll keep you updated with progress there and let you know when that feature hits the development branch.  Thanks for the feature request!

--Madison

**Comment by jmhessel** [comment]
Hi @madisonmay !! Gotcha -- thanks for the heads up -- not sure how I missed that parameter. Does .fit automatically reload the weights from that path?

Thanks again!!

**Comment by jmhessel** [comment]
Thanks for the tips!! :-)
